File: devices-api-java-mvc/src/test/java/com/lucas/devicesapijavamvc/domain/DeviceTest.java
package com.lucas.devicesapijavamvc.domain;

import org.junit.jupiter.api.Test;

import java.time.OffsetDateTime;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

class DeviceTest {

    @Test
    void createNew_shouldGenerateExternalIdAndCreationTime() {
        Device device = Device.createNew("iPhone 15", "Apple", DeviceState.AVAILABLE);

        assertThat(device.getId()).isNull();
        assertThat(device.getExternalId()).isNotNull();
        assertThat(device.getName()).isEqualTo("iPhone 15");
        assertThat(device.getBrand()).isEqualTo("Apple");
        assertThat(device.getState()).isEqualTo(DeviceState.AVAILABLE);
        assertThat(device.getCreationTime()).isNotNull();
    }

    @Test
    void fromEvent_shouldUseAllEventFields() {
        UUID externalId = UUID.randomUUID();
        OffsetDateTime created = OffsetDateTime.now().minusHours(2);

        Device device = Device.fromEvent(
                externalId,
                "Galaxy S24",
                "Samsung",
                DeviceState.IN_USE,
                created
        );

        assertThat(device.getId()).isNull();
        assertThat(device.getExternalId()).isEqualTo(externalId);
        assertThat(device.getName()).isEqualTo("Galaxy S24");
        assertThat(device.getBrand()).isEqualTo("Samsung");
        assertThat(device.getState()).isEqualTo(DeviceState.IN_USE);
        assertThat(device.getCreationTime()).isEqualTo(created);
    }
}

File: devices-api-kotlin-reactive/src/test/kotlin/com/lucas/devicesapikotlinreactive/domain/model/DeviceTest.kt
package com.lucas.devicesapikotlinreactive.domain.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.time.OffsetDateTime
import java.util.UUID

class DeviceTest {

    @Test
    @DisplayName("create() builds a new Device with generated externalId and creationTime")
    fun createShouldBuildNewDevice() {
        val device = Device.create(
            name = "iPhone 15",
            brand = "Apple",
            state = DeviceState.AVAILABLE
        )

        assertThat(device.id).isNull()
        assertThat(device.externalId).isNotNull()
        assertThat(device.name).isEqualTo("iPhone 15")
        assertThat(device.brand).isEqualTo("Apple")
        assertThat(device.state).isEqualTo(DeviceState.AVAILABLE)
        assertThat(device.creationTime).isNotNull()
    }

    @Test
    @DisplayName("fromEvent() builds a Device using the provided event data")
    fun fromEventShouldBuildDeviceFromEventData() {
        val externalId = UUID.randomUUID()
        val creationTime = OffsetDateTime.now().minusDays(1)

        val device = Device.fromEvent(
            externalId = externalId,
            name = "Galaxy S24",
            brand = "Samsung",
            state = DeviceState.IN_USE,
            creationTime = creationTime
        )

        assertThat(device.id).isNull()
        assertThat(device.externalId).isEqualTo(externalId)
        assertThat(device.name).isEqualTo("Galaxy S24")
        assertThat(device.brand).isEqualTo("Samsung")
        assertThat(device.state).isEqualTo(DeviceState.IN_USE)
        assertThat(device.creationTime).isEqualTo(creationTime)
    }
}
